sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+device programs.
+
+For example, in order to implement specialization constants on top of SPIR-V, we
+need to be able to map addresses of `specialization_id` variables into numeric
+IDs of corresponding specialization constants at SPIR-V level.
+
+Another example is device global [implementation][device-global-design], where
+in order to communicate a value of `device_global` variable between host and
+device we need to map its host address to a symbolic name/identifier and some
+other info, which is used at PI layer and below.
+
+This design document describes a generic way how to map address of any SYCL
+object defined in a namespace scope to its unique symbolic ID. Please note that
+this document doesn't try to map the address to something other than a unique
+symbolic ID: other required information is usually generated by the device
+compiler and communicated to the runtime by device image properties. Unique
+symbolic ID which can be obtained from mapping mechanism described in this
+design document could be used as a key in those properties to propagate
+additional information using existing mechanisms.
+
+So, overall the picture looks like:
+- device compiler generates property set/s which provide mapping
+  "unique symbolic ID" -> "various information required by DPC++ RT"
+- device or host compiler generates mapping
+  "address of a host variable" -> "unique symbolic ID" (as described below by
+  this document)
+- DPC++ RT uses these to mappings to obtain required information and somehow
+  uses it
+
+This design document describes two approaches of how the 
+"address of a host variable" -> "unique symbolic ID" mapping can be generated:
+the first one with integration footer and another one with modification of the
+host compiler.
+
+Both approaches have their pros and cons and they are expected to be implemented
+and exists in the implementation at the same time, but only one of them will be
+used at a time depending on whether 3rd-party host compiler is used or not.
+
+Integration footer can be used with 3rd-party host compilers, but it requires
+appending to a translation unit provided by user, which could affect debug
+information: since there are no compilers that support appending a file at the
+end (similar to `-include`), appending is done by generating a temporary input
+file using concatenation of the original input and integration footer.
+
+Such replacement of the main translation unit causes the following issues:
+- debug information about the source file might be incorrect, leading to
+  problems with gdb `l` command and code coverage tools
+- checksum of host and device source files becomes different which causes device
+  code debugging to be completely broken in some environments (such as MS Visual
+  Studio, for example)
+
+Customizing host compiler allows to avoid issues with debuggers and code
+coverage tools, but that is not an option if user wants to compile host part
```

Comment:
```suggestion
coverage tools, but that is not an option if a user wants to compile host part
```

Reply:
Applied in 6c7a9be71e57c2855e87e18de145a833629dc060